enhancements/single-node-production-deployment-approach.md

```diff
@@ -0,0 +1,511 @@
+teams developing OpenShift components will need to consider how their
+components should be configured when deployed and used when the
+high-availability capability is disabled.
+
+In the telco RAN use case, high-availability is typically achieved by
+having multiple sites provide coverage to overlapping geographic
+areas. Therefore, use of cluster-based high-availability features will
+be limited (for example, by running a single API service).
+
+Failures in edge deployments are frequently resolved by re-imaging or
+physically replacing the entire host. Combining this fact with the
+previous observation about the approach to providing highly-available
+services lets us draw the conclusion that in-place upgrades do not
+need to be supported by the first iteration of this configuration.
+
+In a multi-node cluster, operators may safely trigger rolling reboots
+or restarts of the control plane services without taking the cluster
+offline. In a single-node deployment, any such restart would take some
+or all of the services offline. The deployment or upgrade of a cluster
+may include several host reboots, clustered together in time, until
+the operators stabilize. Users should wait for this period of
+instability before installing their workloads.
+
+The single-node configuration of OpenShift will be sufficiently
+generic to cater to a variety of edge computing use cases. As such,
+OpenShift's usual cluster configuration mechanisms will be favored
+where there is a likelihood that there will be edge computing use case
+with differing requirements.  For example, there is no expectation
+that single-node deployments will use the real-time kernel by
+default -- this will continue to be a `MachineConfig` choice as per
+[enhancements/support-for-realtime-kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md). *See open questions*
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+deployment, with some limitations (see non-goals above). The user
+will not require special support exceptions to receive technical assistance
+for the features supported by the configuration.
+
+### Implementation Details/Notes/Constraints
+
+Some OpenShift components require a minimum of 2 or 3 nodes. The new
+cluster capabilities API will be used to configure these components as
+appropriate for a single node.
+
+#### installer
+
+We will add logic to the installer to set the new
+`highAvailabilityMode` status field of the `infrastructure` API
+resource. The will value be set to `Full` by default and `None` when
+the number of replicas for the control plane is `1`. See the [cluster
+high-availability mode
+API](https://github.com/openshift/enhancements/pull/555) enhancement
+for details.
+
+#### machine-config-operator
+
+The machine-config-operator includes a check in code that the number
```

Comment:
I think the initial prototype testing code here may have led to an incorrect assumption.

If e.g. the RT kernel is provided as [a day 0 manifest](https://github.com/openshift/installer/blob/master/docs/user/customization.md#install-time-customization-for-machine-configuration) it will end up in the Ignition which is used for the "Ignition boot" and most importantly *will be applied before kubelet is up*. 

So if we're not supporting in place upgrades *and* we're also not supporting "day 2 machineconfig" changes, then it's not clear to me that the MCO needs to be patched at all (though we need to be sure it doesn't go degraded etc.)

Is there a need for "day 2 machineconfig" changes in this? 

(That said *personally* I don't think the MCO changes to support single-node upgrades by skipping drain would be very hard, and as long as we have CI it seems pretty supportable)

Reply:
There is a need  for day2 machine config to support PAO / NTO operators. @SchSeba and @MarSik can elaborate.

Reply:
Hmm oh wait part of the problem here is we want to support https://github.com/openshift-kni/performance-addon-operators/ which generate MachineConfig and may even have per-node configuration?

If we can avoid other components generating MachineConfig dynamically in this scenario - or to say it positively: if the MachineConfigs generated by performance-addons (for things like kernel arguments) can be determined "statically" and passed to the installer, then we can avoid another reboot and a need to change the MCO.

Reply:
we had this discussions with them. It will mean to move the operator logic to the 'user'....

Reply:
A possible alternative is to run PAO/NTO during bootstrap and add the MachineConfigs they generate to the manifests that form the master Ignition.

Reply:
Nice :) We should get the team input on that

Reply:
This work does sound like it would be useful, but I think it is worth writing up as a separate enhancement as an optimization on top of this first iteration.